src/hotspot/share/opto/rangeinference.cpp

```diff
@@ -0,0 +1,559 @@
+/*
+ * Copyright (c) 2024, Oracle and/or its affiliates. All rights reserved.
+ * DO NOT ALTER OR REMOVE COPYRIGHT NOTICES OR THIS FILE HEADER.
+ *
+ * This code is free software; you can redistribute it and/or modify it
+ * under the terms of the GNU General Public License version 2 only, as
+ * published by the Free Software Foundation.
+ *
+ * This code is distributed in the hope that it will be useful, but WITHOUT
+ * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
+ * FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License
+ * version 2 for more details (a copy is included in the LICENSE file that
+ * accompanied this code).
+ *
+ * You should have received a copy of the GNU General Public License version
+ * 2 along with this work; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301 USA.
+ *
+ * Please contact Oracle, 500 Oracle Parkway, Redwood Shores, CA 94065 USA
+ * or visit www.oracle.com if you need additional information or have any
+ * questions.
+ *
+ */
+
+#include "precompiled.hpp"
+#include "opto/rangeinference.hpp"
+#include "opto/type.hpp"
+#include "utilities/tuple.hpp"
+
+constexpr juint SMALLINT = 3; // a value too insignificant to consider widening
+
+template <class T>
+class AdjustResult {
+public:
+  bool _progress;
+  bool _present;
+  T _data;
+};
+
+template <class T>
+class NormalizeSimpleResult {
+public:
+  bool _present;
+  RangeInt<T> _bounds;
+  KnownBits<T> _bits;
+};
+
+// Try to tighten the bound constraints from the known bit information
+// E.g: if lo = 0 but the lowest bit is always 1 then we can tighten
+// lo = 1
+template <class T>
+static AdjustResult<RangeInt<T>>
+adjust_bounds_from_bits(const RangeInt<T>& bounds, const KnownBits<T>& bits) {
+  static_assert(std::is_unsigned<T>::value, "");
+
+  auto adjust_lo = [](T lo, const KnownBits<T>& bits) {
```

Comment:
Since it is only used here I think it would be more sensible to make it a local lambda to lower the visibility, the resulting function is not too large, too.

Reply:
This frightens me ...
40 lines of code is not small. What benefit it has vs normal private method?